core/src/main/java/org/apache/druid/timeline/SegmentWithOvershadowedStatus.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.timeline;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+/**
+ * DataSegment object plus the overshadowed status for the segment. An immutable object.
+ *
+ * SegmentWithOvershadowedStatus's {@link #compareTo} method considers only the {@link SegmentId} of the DataSegment object.
+ */
+public class SegmentWithOvershadowedStatus implements Comparable<SegmentWithOvershadowedStatus>
```

Comment:
Did you explore the possibility for this class to extend `DataSegment` for memory saving purposes?

Reply:
I don't see why `SegmentWithOvershadowedStatus` should have a "dataSegment" field rather than all fields deconstructed. In fact, it would allow saving a little of serialization/deserialization and the number of bytes sent over the network as well.

Reply:
If I deconstruct the `DataSegment` object, then we might save few bytes on a reference to DataSegment, but then the memory savings in broker where interned DataSegment is used would be lost. (Getting rid of interning or not is another issue, which should be addressed outside of this PR). If the concern is bytes sent over the network, then moving to smile format instead of json can provide considerable reduction in size of bytes transferred, which I plan to do in a follow-up PR later.

Reply:
Is `SegmentWithOvershadowedStatus` stored somewhere for a long time? Intering an object upon deserialization and throwing it away soon doesn't make a lot of sense.

And even if the overshadowed status should be kept around on some node for long time, you would better off apply mapping techniques such as described in #7395 instead of using plain Guava's interners. When you do this, you can insert the "overshadowed" flag wherever you want, or have something like `ConcurrentHashMap<DataSegment, SegmentWithOvershadowedStatus>` for storage, etc.

Reply:
> If I deconstruct the DataSegment object, then we might save few bytes on a reference to DataSegment, but then the memory savings in broker where interned DataSegment is used would be lost. 

With a flat representation in the wire format, you could still extract->create a DataSegment object and intern that.

That said, given that there's a plan to move to smile format, I would lean towards keeping the existing SegmentWithOvershadowedStatus with the separate DataSegment field, since I like that structure better. With smile, the extra reference would only add an extra dictionary-encoded key and two braces, I believe, so the byte savings would be minimal and dwarfed by things like the column name list.

I think this area is good enough for the goals of this PR, and we can always adjust this area later if needed as part of resolving the interning discussion in #7395

Reply:
> Is SegmentWithOvershadowedStatus stored somewhere for a long time? Intering an object upon deserialization and throwing it away soon doesn't make a lot of sense.

It's stored long-term in the Druid SQL segment metadata cache. Which we may want to unify at some point with other segment caches in the Broker (#7395). But for the context of this patch I don't think this needs to be done. It's mixing too many concerns into the same patch.

Reply:
If this PR gets into the next Druid release but any future PRs that cancel this change won't (as well as smile format change), it wouldn't be easy to get rid of the current APIs, because `SegmentWithOvershadowedStatus` is transmitted between different types of Druid nodes which can be updated separately.

Reply:
The coordinator API is evolvable, and is already evolving in this patch via request parameters: its response structure is different based on whether or not the `includeOvershadowedStatus` parameter is provided. If it needs to evolve further, then that would be okay and doable. (Although if all we do is switch to Smile, I don't think structural evolution is needed, since I imagine we would do that switch by making various APIs support both JSON and Smile based on a client header.)

By the way, we could get rid of the old formats after a few releases if we want, by deprecating them and then introducing a hard barrier that rolling updates cannot cross. We usually try to avoid doing this too often but it can be done.

Reply:
BTW, I think we should have something like `@ClusterInternalAPI` annotation for this.

Reply:
sounds like a good idea to add an annotation for all internal API's in a separate PR

Reply:
https://github.com/apache/incubator-druid/issues/7542

Reply:
thanks